Add a non-static API for the CssSelector component

stof opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #15850, #8404 |
| License | MIT |
| Doc PR | todo |

This implements a non-static API for the CssSelector component.

I decided to keep the static API too, as it is convenient when you just need a one-shot conversion (if you need lots of conversions, keeping a reference to the Converter and all its internal object graph may be faster than releasing it all the time and rebuilding it).
I deprecated the global state to choose between HTML and XML conversion. The static API would always enable the HTML extension in 3.0. Dealing with XML would be done by using the Converter class.

A second commit also tags all internal classes of the component as `@internal`, as there is really no reason for a user to deal with them (btw, we already considered them fully internal in the past, as we broke BC on them in a patch release to fix memory performance of the component in the past).

TODOs:
- [x] Validate whether we keep the static facade to the component
- [ ] send a PR on the documentation to document this new API.
- [x]  handle usage of the deprecated API in the DomCrawler testsuite

The DomCrawler component does not use the new API yet. I will do it in a separate PR, as distinguishing between HTML and XML modes for a crawler will be easier once I deprecate the possibility to load multiple documents (which I will do tomorrow).

stof committed:
Add a non-static API for the CssSelector component

stof committed:
[CssSelector] Tag all internal classes as internal ones

All these classes are meant to be considered as an implementation
detail. A normal usage of the component does not require to deal with
them at all.

stof commented:
@symfony/deciders please give your mind on this

jakzal commented:
Looks good to me :+1: 

I would get rid of the static API as I prefer to have one way of doing the same thing. `CssSelector::toXpath('h2');` vs  `(new Converter())->toXpath('h2');` is not much of a difference.

xabbuh commented on src/Symfony/Component/CssSelector/ConverterInterface.php:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @return string
+     *
+     * @api
```
Why not tagging the whole interface instead?

Tobion replied:
should be removed. we're not tagging from the get-go

stof replied:
@Tobion IMO, it makes sense to tag it now, as this is the API we expect them to use (or we should drop this tagging altogether).

Tobion replied:
the api tag just means it has stricter BC rules. it doesn't make sense to enforce that on a new class without usage experience. otherwise we could add it everywhere.

stof replied:
Well, we haven't tagged anything as ``@api`` since 2.3 AFAICT. And we still work hard on BC for all these classes.

Btw, the api tag BC rules are not stricter. they are the real rules. the special rules are for non-api classes, where we allow BC breaks, but only some of them. And as we became better at writing BC layers and worse at tagging things as ``@api``, this distinction does not really make sense IMO. We already apply strict BC almost everywhere

Tobion replied:
I agree with what you say. Maybe it's time to remove the api tags then.

stof replied:
see https://github.com/symfony/symfony/issues/15977

xabbuh commented on src/Symfony/Component/CssSelector/CssSelector.php:
```diff
@@ -82,35 +75,32 @@ class CssSelector
      */
     public static function enableHtmlExtension()
     {
+        @trigger_error('The method '.__METHOD__.' is deprecated as of 2.8 and will be removed in 3.0. Use \Symfony\Component\CssSelector\Converter instead.', E_USER_DEPRECATED);
```
Usually we wrote `The '.__METHOD__.' method is [...]'`.

xabbuh replied:
And I would change "Use \Symfony\Component\CssSelector\Converter instead." to "Use the \Symfony\Component\CssSelector\Converter class instead." (same as in the docblock).

stof replied:
Ah, I forgot this place

xabbuh commented:
Looks good to me and I agree with @jakzal that we can get rid of the static API in 3.0.

fabpot commented:
:+1: for removing the static API as well in 3.0.

fabpot commented on src/Symfony/Component/CssSelector/Converter.php:
```diff
@@ -0,0 +1,52 @@
+    {
+        return $this->translator->cssToXPath($cssExpr, $prefix);
+    }
+
```
blank line should be removed

Tobion commented on src/Symfony/Component/CssSelector/ConverterInterface.php:
```diff
@@ -0,0 +1,31 @@
+
+namespace Symfony\Component\CssSelector;
+
+interface ConverterInterface
```
what do we need to Interface for? Will there be different implementations?

Tobion commented on src/Symfony/Component/CssSelector/Converter.php:
```diff
@@ -0,0 +1,52 @@
+use Symfony\Component\CssSelector\XPath\Extension\HtmlExtension;
+use Symfony\Component\CssSelector\XPath\Translator;
+
+class Converter implements ConverterInterface
```
Imo the class name is to generic. How about `CssToXPathConverter`?

Tobion commented:
:+1: for deprecating the static method

stof commented:
The PR is updated.

As the static API is now deprecated, I switched the DomCrawler component to use the non-static API already. Currently, it switched to XML mode as soon as you add XML content (and is HTML by default), which closes https://github.com/symfony/symfony/issues/8404
As soon as #15849 is done (this evening), we will not have the possibility to load both XML and HTML content in the same crawler, making this unambiguous.

I extracted the creation of sub crawlers (created by filtering an existing one) to a dedicated private method to avoid duplication of the code copying the state in multiple places in Symfony. this is something I would have done anyway as I will have more state to copy when doing the DomCrawler changes. This is cleaner than adding new constructor arguments used only internally (it would clutter the public API). Btw, this allowed me to spot a place where the baseHref property was not passed to the new crawler properly.

Tobion commented on src/Symfony/Component/CssSelector/CssSelector.php:
```diff
@@ -82,24 +79,15 @@ class CssSelector
     /**
      * Enables the HTML extension.
+     *
+     * @deprecated as of 2.8, will be removed in 3.0. Use the \Symfony\Component\CssSelector\Converter class instead.
```
the class is already deprecated

stof replied:
fixed

xabbuh commented on src/Symfony/Component/CssSelector/ConverterInterface.php:
```diff
@@ -0,0 +1,75 @@
+ *
+ * @author Christophe Coevoet <stof@notk.org>
+ *
+ * @api
```
should be removed now

xabbuh replied:
same with the tag on the method of course

fabpot commented:
Thank you @stof.

xabbuh commented on src/Symfony/Component/DomCrawler/composer.json:
```diff
@@ -20,7 +20,7 @@
+        "symfony/css-selector": "~2.8|~3.0.0"
     },
     "suggest": {
         "symfony/css-selector": ""
```
Should we expand this to explain that we need the CssSelector component in version 2.8 or higher?

xabbuh commented:
> I extracted the creation of sub crawlers (created by filtering an existing one) to a dedicated private method to avoid duplication of the code copying the state in multiple places in Symfony. this is something I would have done anyway as I will have more state to copy when doing the DomCrawler changes. This is cleaner than adding new constructor arguments used only internally (it would clutter the public API). Btw, this allowed me to spot a place where the baseHref property was not passed to the new crawler properly. 

What about backporting this change to Symfony 2.3? At least, we should fix the `slice()` method there.

Tobion commented:
By comments have no been taken into account:

> what do we need to Interface for? Will there be different implementations?

> Imo the class name is too generic. How about CssToXPathConverter?

fabpot commented:
I did not read your comment about the interface; indeed, I'm in favor of removing this interface as I don't see how we could have another implementation.

fabpot commented:
@Tobion see #15988 for the interface removal.
